### PR TITLE
improvements for sub

### PR DESCRIPTION
**Pull request by drtrigon**
improvements;
* make sub error handling working
* make sure element size also depends on number of in/outputs

### CODE REVIEW

**Comment by drtrigon** [comment]
improvments; * add single step capability (do one step by right-click)

**Comment by drtrigon** [comment]
improvements; * prevent sub window close (make it borderless)

This might not be optimal but it makes sure that the window do not get closed and by that invalid.

**Comment by drtrigon** [comment]
Possible solution for https://github.com/honix/Pyno/issues/9

**Comment by drtrigon** [comment]
Seems to need a bugfix as json.loads return lists instead of tuples.

**Comment by honix** [comment]
Hey! I think 1d88f2f is little dissappointing, i cant drag sub window around using borderless style. 

Using Windows 10.

File: menu.py [hunk 1]
```diff
@@ -14,42 +16,51 @@ def copy_nodes(window, data=False):
...
 def paste_nodes(window, data=None):
     x, y = (0, 0) if data else (window.pointer[0], window.pointer[1])
     buff = []
     try:
-        paste = eval(data) if data else eval(pyperclip.paste())
+        #paste = eval(data) if data else eval(pyperclip.paste())
+        data = data if data else pyperclip.paste()
+        try:
+            paste = json.loads(data)
+            #paste = json.loads(data, object_pairs_hook=OrderedDict)
```
Comment (bilderbuchi) [inline menu.py]: Why not use the object_pairs_hook variant, I think this is the canonical way to do it?

**Comment by bilderbuchi** [comment]
> json.loads return lists instead of tuples.
Does it matter if your return is a list instead tuple (except less "safety" as tuples are immutable)?

File: menu.py (hunk 1 above)
Comment (bilderbuchi) [reply menu.py]: btw, great work on implementing this!

Comment (drtrigon) [reply menu.py]: May I just don't understand this, but I don't see the point in getting an OrderedDict when loading the data, it's of no use. I see the point in saving, in order to have a well ordered text/json file for diffs e.g. - and that is implemented.

**Comment by drtrigon** [comment]
> Hey! I think [1d88f2f](https://github.com/honix/Pyno/commit/1d88f2f84f2416133b8242311706f494c9b5e261) is little dissappointing, i cant drag sub window around using borderless style.
> 
> Using Windows 10.

Yeah I totally agree with you it is ugly and hacky (that's why this commit was the last one). The point is we HAVE TO prevent the user from closing windows. I was not able to figure out how to make pyglet to disable the window close button. The only other way I found is just to ignore the close button, but that involves quite some changes to `on_close` as it needs to be able to figure out what is the root/main window and allow to close that only. So if you have a better approach? I would be happy with it!

**Comment by drtrigon** [comment]
> > json.loads return lists instead of tuples.
> 
> Does it matter if your return is a list instead tuple (except less "safety" as tuples are immutable)?
> I think your workaround is fine - at the interface to the data at rest/in the file, this makes clear where you need tuples instead of lists.

Yes it does matter, because of this:
```
>>> a = (0, 1)
>>> a + (3,)
(0, 1, 3)
>>> a = [0, 1]
>>> a + (3,)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can only concatenate list (not "tuple") to list
```

**Comment by honix** [comment]
@drtrigon i will merge it with borderless subwindows, maybe later better solution pops up.

**Comment by drtrigon** [comment]
> @drtrigon i will merge it with borderless subwindows, maybe later better solution pops up.

Was exactly my thinking... ;)

**Pull request merged**